File: img2color.py
#!/usr/bin/env python2

from PIL import Image
import sys
import random
import argparse

def color_diff(a, b):
	d0 = a[0]-b[0]
	d1 = a[1]-b[1]
	d2 = a[2]-b[2]
	d0 = d0 if d0>0 else d0*-1
	d1 = d1 if d1>0 else d1*-1
	d2 = d2 if d2>0 else d2*-1
	return d0+d1+d2

def med(cluster, k):
	if(len(cluster) == 0):
		return k
	r,g,b = (0,0,0)
	for c in cluster:
		r = r + c[0]
		g = g + c[1]
		b = b + c[2]
	r = r/len(cluster)
	g = g/len(cluster)
	b = b/len(cluster)
	return [r,g,b]


def assign_cluster(im, k, cluster, width, height):
	minimum = len(k)
	for i in range(0, width):
		for j in range(0, height):
			minimum = len(k)
			difference = 766
			for l in range(0, len(k)):
				new_diff = color_diff(im.getpixel((i, j)), k[l])
				if(new_diff < difference):
					difference = new_diff
					minimum = l
			cluster[minimum].append(im.getpixel((i, j)))

def kmeans(im):
	width, height = im.size
	k = []
	cluster = []
	cluster1 = []
	cluster2 = []
	cluster3 = []
	for i in range(0,3):
		k.append(im.getpixel((random.randint(0,width), random.randint(0,height))))
		cluster.append([])
		cluster1.append([])
		cluster2.append([])
		cluster3.append([])
	print k
	assign_cluster(im, k, cluster, width, height)
	k[0] = med(cluster[0], k[0])
	k[1] = med(cluster[1], k[1])
	k[2] = med(cluster[2], k[2])
	print k
	assign_cluster(im, k, cluster1, width, height)
	k[0] = med(cluster1[0], k[0])
	k[1] = med(cluster1[1], k[1])
	k[2] = med(cluster1[2], k[2])
	print k
	assign_cluster(im, k, cluster2, width, height)
	k[0] = med(cluster2[0], k[0])
	k[1] = med(cluster2[1], k[1])
	k[2] = med(cluster2[2], k[2])
	print k
	assign_cluster(im, k, cluster3, width, height)
	k[0] = med(cluster3[0], k[0])
	k[1] = med(cluster3[1], k[1])
	k[2] = med(cluster3[2], k[2])
	print k
	return k



def main():
	parser = argparse.ArgumentParser(description='Find main colors in a given image.')
	parser.add_argument("image", help="Image to be processed")
	parser.add_argument("--output-format",type=str, choices=['image-palette', 'html-color-code'], help="Output-format")
	args = parser.parse_args()
	print args
	image = Image.open(args.image)
	k = kmeans(image)
	if(args.output_format == "image-palette"):
		img = Image.new('RGB', (image.size[0]+100, image.size[1]))
		img.paste(image)
		img0 = Image.new('RGB', (100, image.size[1]/3), color = tuple(k[0]))
		img1 = Image.new('RGB', (100, image.size[1]/3), color = tuple(k[1]))
		img2 = Image.new('RGB', (100, image.size[1]/3), color = tuple(k[2]))
		
		img.paste(img0, box=(image.size[0], 0))
		img.paste(img1, box=(image.size[0], image.size[1]/3))
		img.paste(img2, box=(image.size[0], 2*image.size[1]/3))

		img.save(args.image +"_pallette.png", "PNG")	
	else:
		print "#%02x%02x%02x" % (k[0][0],k[0][1],k[0][2])
		print "#%02x%02x%02x" % (k[1][0],k[1][1],k[1][2])
		print "#%02x%02x%02x" % (k[2][0],k[2][1],k[2][2])



if __name__ == "__main__":
	main()

File: img2color.go
package main

import (
	"flag"
	"fmt"
	"image"
	"image/color"
	_ "image/gif"
	_ "image/jpeg"
	"image/png"
	"log"
	"math"
	"math/rand"
	"os"
	"strings"
)

// color_diff_euklid is a function to calculate the distance between to colors.
// The parameters are a: a RGB-array with 16bit color values and b: a RGB-array with 8bit color values.
// It returns the euklidian distance between the colors.
func color_diff_euklid(a [3]int, b []int) int {
	return int(math.Sqrt(math.Pow(float64((a[0]>>8)-b[0]), 2) + math.Pow(float64((a[1]>>8)-b[1]), 2) + math.Pow(float64((a[2]>>8)-b[2]), 2)))
}

// assign_k is a function to assign each pixel of the image to one k.
// Since it is to be run multithreaded, the result is returned via a channeldoc: symbol img2color is not a type in package main installed in "."
// The start and stop of the assigned pixels are returned in the last field of the result matrix tmp_mat
// This is done to achieve deterministic results
func assign_k(image image.Image, k int, k_med [][]int, start int, stop int, ch chan [][]int) {
	tmp_mat := make([][]int, stop-start+1)
	for i := 0; i < len(tmp_mat); i++ {
		tmp_mat[i] = make([]int, image.Bounds().Max.Y)
	}
	for x := start; x < stop; x++ {
		for y := 0; y < (image).Bounds().Max.Y; y++ {
			minimum := k
			difference := 766
			for i := 0; i < k; i++ {
				R, G, B, _ := image.At(x, y).RGBA()
				new_diff := color_diff_euklid([3]int{int(R), int(G), int(B)}, k_med[i])
				if new_diff < difference {
					difference = new_diff
					minimum = i
				}
			}
			tmp_mat[x-start][y] = minimum
		}
	}
	tmp_mat[len(tmp_mat)-1][0] = start
	tmp_mat[len(tmp_mat)-1][1] = stop
	ch <- tmp_mat
}

// medium_k is used to calculate the medium color of all pixels that are assigned to one k
// It is used multithraded and the color is returned via a channel
func medium_k(image image.Image, j int, width int, height int, k_mat [][]int, ch_m chan []int) {
	k_m := make([]int, 4)
	count := 0
	for x := 0; x < width; x++ {
		for y := 0; y < height; y++ {
			if k_mat[x][y] == j {
				count++
				r, g, b, _ := image.At(x, y).RGBA()
				k_m[0] += int(r) >> 8
				k_m[1] += int(g) >> 8
				k_m[2] += int(b) >> 8
			}
		}
	}
	if count > 0 {
		k_m[0] /= count
		k_m[1] /= count
		k_m[2] /= count
	}
	k_m[3] = j
	ch_m <- k_m
}

// kmeans is the function to calculate k mean colors
// image is the image to be processed
// k is the number of mean colors to be calculated
// t is the number of threads to be used
// n is the number of rounds the algorithm shall run. higher n achieves better results, but from a certain n up the mean colors will not change anymore
func kmeans(image image.Image, k int, t int, n int) [][]int {
	ch := make(chan [][]int)
	ch_m := make(chan []int)

	k_med := make([][]int, k)
	r, _, _, _ := image.At(0, 0).RGBA()
	rand.Seed(int64(r >> 8))
	for i := 0; i < k; i++ {
		R, G, B, _ := image.At(int(rand.Int31n(int32(image.Bounds().Max.X))), int(rand.Int31n(int32(image.Bounds().Max.Y)))).RGBA()
		k_med[i] = []int{int(R) >> 8, int(G) >> 8, int(B) >> 8}
	}
	k_mat := make([][]int, image.Bounds().Max.X)
	for i := 0; i < len(k_mat); i++ {
		k_mat[i] = make([]int, image.Bounds().Max.Y)
	}
	for i := 0; i < n; i++ {
		for j := 0; j < t; j++ {
			start := int(math.Round(float64(j) * float64(image.Bounds().Max.X) / float64(t)))
			stop := int(math.Round((float64(j) + 1) * float64(image.Bounds().Max.X) / float64(t)))
			go assign_k(image, k, k_med, start, stop, ch)
		}
		for j := 0; j < t; j++ {
			re := <-ch
			copy(k_mat[re[len(re)-1][0]:re[len(re)-1][1]], re[0:len(re)-1])
		}
		for j := 0; j < k; j++ {
			go medium_k(image, j, image.Bounds().Max.X, image.Bounds().Max.Y, k_mat, ch_m)
		}
		for j := 0; j < k; j++ {
			k_m := <-ch_m
			k_med[k_m[3]] = k_m[:3]
		}

		fmt.Printf("\rProcessing:\t%.2f%%", (float64(i)*100)/float64(n))
	}
	fmt.Printf("\rProcessing:\t100.00%%")
	fmt.Println("\nDone.")
	return k_med
}

// main is used to interpret the parameters,
// start the algorithm and output the results
func main() {
	k_ptr := flag.Int("k", 5, "Number of colors to find")
	t_ptr := flag.Int("t", 1, "Number of threads to use for computation")
	n_ptr := flag.Int("n", 10, "Number of rounds for computation")
	//	fast_ptr := flag.Bool("fast", false, "Activate fast mode.")
	var image_file string
	flag.StringVar(&image_file, "image", "", "Image to be processed")
	var output_ptr string
	flag.StringVar(&output_ptr, "output", "palette", "Output option")
	flag.Parse()

	reader, err := os.Open(image_file)
	if err != nil {
		log.Fatal(err)
	}
	m, _, err := image.Decode(reader)
	if err != nil {
		log.Fatal(err)
	}

	k_med := kmeans(m, *k_ptr, *t_ptr, *n_ptr)

	for i := 0; i < len(k_med); i++ {
		fmt.Printf("#%02x%02x%02x\n", k_med[i][0], k_med[i][1], k_med[i][2])
	}

	fmt.Println(output_ptr)

	if strings.Compare(output_ptr, "silhouette") == 0 {
		img := image.NewRGBA(image.Rectangle{image.Point{0, 0}, image.Point{m.Bounds().Max.X, m.Bounds().Max.Y}})
		for x := 0; x < m.Bounds().Max.X; x++ {
			for y := 0; y < m.Bounds().Max.Y; y++ {
				minimum := *k_ptr
				difference := 766
				for i := 0; i < *k_ptr; i++ {
					R, G, B, _ := m.At(x, y).RGBA()
					new_diff := color_diff_euklid([3]int{int(R), int(G), int(B)}, k_med[i])
					if new_diff < difference {
						difference = new_diff
						minimum = i
					}
				}
				img.Set(x, y, color.RGBA{uint8(k_med[minimum][0]), uint8(k_med[minimum][1]), uint8(k_med[minimum][2]), 0xff})
			}
		}
		f, _ := os.Create("image_s.png")
		png.Encode(f, img)
	}

	if strings.Compare(output_ptr, "palette") == 0 {
		img := image.NewRGBA(image.Rectangle{image.Point{0, 0}, image.Point{m.Bounds().Max.X + 100, m.Bounds().Max.Y}})
		for x := 0; x < m.Bounds().Max.X; x++ {
			for y := 0; y < m.Bounds().Max.Y; y++ {
				img.Set(x, y, m.At(x, y))
			}
		}
		k := *k_ptr
		for x := m.Bounds().Max.X; x < m.Bounds().Max.X+100; x++ {
			for y := 0; y < m.Bounds().Max.Y; y++ {
				img.Set(x, y, color.RGBA{uint8(k_med[(y*k)/m.Bounds().Max.Y][0]), uint8(k_med[(y*k)/m.Bounds().Max.Y][1]), uint8(k_med[(y*k)/m.Bounds().Max.Y][2]), 0xff})
			}
		}
		f, _ := os.Create("image_p.png")
		png.Encode(f, img)
	}

	defer reader.Close()
}
